lib/haml_lint/tree/tag_node.rb

```diff
@@ -7,9 +7,20 @@ class TagNode < Node
     # different attribute setting syntaxes (`{...}`/`(...)`), converted into
     # Ruby code.
     #
+    # @note This has to be memoized because of a design decision in Haml 5. When
+    # calling `DynamicAttributes#to_literal`, they mutate the "old" parameter using
+    # `String#sub!` instead of returning a new string. This means that any subsequent
+    # calls can return a nil instead of a string for that attribute, which causes
+    # any subsequent calls to the method to raise an error.
+    #
     # @return [Array<String>]
     def dynamic_attributes_sources
-      @value[:attributes_hashes]
+      @dynamic_attributes_sources ||=
+        if Gem::Version.new(Haml::VERSION).approximate_recommendation == '~> 5.0'
+          Array(@value[:dynamic_attributes].to_literal).reject(&:empty?)
+        else
+          @value[:attributes_hashes]
+        end
```

Comment:
Would it make more sense to structure this to check if we are running Haml < 5 instead? The current check will fail should Haml 6 ever be released, but it's probable that the fix to get this working on Haml 5 will continue to work on Haml 6.

Designing the version check to be based on the information we have about Haml right now seems appropriate.

```ruby
@dynamic_attributes_source ||=
  if Gem::Version.new(Haml::VERSION) < Gem::Version.new('5')
    @value[:attributes_hashes]
  else
    Array(@value[:dynamic_attributes].to_literal).reject(&:empty?)
  end
```